### PR TITLE
[2.x] Add user and shell configuration

### LINKED ISSUE #245: Localhost shell
- Envoy Version: 2.8.2
- PHP Version: 8.0.17

### Description:
The localhost run doesn't use the same shell as the one it's executed on

Consider the following `Envoy.blade.php`
```blade
@servers(['local' => '127.0.0.1', 'dev' => 'root@example.com'])

@task('test')
shopt -s extglob
ls !(git)
@endtask
```

The script will work fine on `root@example.com` because the shell is `bash` (assuming that's the configured shell of the root user)
But it will not work on the local env because it's executed with `sh` even though my shell is configured to be /bin/bash

Also note that because of the usage of `shopt` it's not possible to rewrite this script as `bash -c "shopt -s extglob && ls !(git)"` which would produce a syntax error because shopt hasn't applied yet.

### Steps To Reproduce:
Simply `composer require --dev laravel/envoy` and paste the blade file
```blade
@servers(['local' => '127.0.0.1'])

@task('test')
shopt -s extglob
ls !(git)
@endtask
```

Run ` echo "$SHELL" && vendor/bin/envoy run test`

```txt
/bin/bash
[127.0.0.1]:  sh: 1: shopt: not found
[127.0.0.1]:  sh: 2: Syntax error: "(" unexpected
[✗] This task did not complete successfully on one of your servers.
```

Notice how `$SHELL` is `/bin/bash` but envoy uses `/bin/sh`

### Resolution

Ideally envoy should get the name of the shell in which it is running and run the local command using that same shell, though php doesn't seem to give the option with `proc_exec` so this would require a workaround of spawning a shell within proc_exec

A nice addition when this is fixed would be if we could specify the shell to use in the task arguments as well

### PR DESCRIPTION
**Pull request by Tofandel**
This PR fixes #245, because it would run on bash on ssh (see bash -e in the code) but not on localhost
Some scripts would then behave differently when run on localhost, this defaults to bash on localhost as well and adds the options to change the shell in both ssh and localhost

During development I also noticed that the `as` option had been added since the beggining (see https://github.com/laravel/envoy/blame/1eb1e129bd3eca5513928352cfd4368afbb12c57/src/Task.php#L17) but never actually used anywhere, so this PR also adds this features

# Features

- [x] Run as specific user on linux 
- [ ] Run as specific user on windows (feasible with `runas /user:administrator` but will ask for user password, not the case with sudo NOPASSWD, might need separate option for password input)
- [x] Run in a specific shell on linux  
- [ ] Run in a specific shell on windows (powershell or cmd)
- [ ] Detect correct implementation based on remote OS and requested shell (Currently it's assumed that the remote ssh runs the same OS as the script running envoy, it might not be the case)

# Tests
- [x] Build a E2E test framework
- [x] Tests for added features
- [ ] Other tests for existing features 
- [ ] Tests for different environment and with ssh (windows, linux etc)

### CODE REVIEW

File: src/Task.php [hunk 1]
```diff
@@ -44,14 +51,16 @@ class Task
...
      * @param  array  $hosts
      * @param  string  $user
+     * @param  string  $shell
      * @param  string  $script
      * @param  bool  $parallel
      * @param  string|null  $confirm
      * @return void
      */
-    public function __construct(array $hosts, $user, $script, $parallel = false, $confirm = null)
+    public function __construct(array $hosts, $user, $shell, $script, $parallel = false, $confirm = null)
```
Comment (driesvints) [inline src/Task.php]: This is a breaking change.

File: src/RemoteProcessor.php [hunk 2]
```diff
@@ -29,16 +29,24 @@ protected function getProcess($host, Task $task)
 
         $env = $this->getEnvironment($host);
 
+        $delimiter = 'EOF-LARAVEL-ENVOY';
+
+        $shell = $task->shell ?: 'bash';
+
+        if (! empty($task->user)) {
+            $shell = 'sudo -u '.escapeshellarg($task->user).' '.$shell;
```
Comment (driesvints) [inline src/RemoteProcessor.php:37]: I'm not so sure about this sudo call? Why is this needed?

**Comment by driesvints** [comment]
PR looks good overall to me. Left a couple of comments.

**Comment by Tofandel** [comment]
Thanks @driesvints, breaking change should be fixed, and commented on the rest, it's ready for review as is and I'll probably work on the missing todos in a bit later this week

**Comment by driesvints** [comment]
Closing this because of inactivity. Feel free to resend this in once all todos have been done.
